### PR TITLE
feat(scope-budget): detect multi-subsystem, multi-issue PRs

### LINKED ISSUE #2067: Extend pr-scope-budget with subsystem-group and multi-issue detection
## Problem

`scripts/pr-scope-budget.mjs` measures changed core lines with thresholds `warnLines: 1500` / `failLines: 4000`. This catches large PRs but cannot detect **scope mixing** — bundling unrelated work (e.g., a regression test for one issue with a feature for a different issue) into a single PR.

Evidence: a recent PR titled as a regression test for one issue also carried unrelated feature work for a different issue. The unrelated feature attracted its own review findings, which then blocked the entire PR including the regression test. Line-count budgeting alone would not have flagged this.

## Proposed solution

Add a subsystem-group classifier to `scripts/pr-scope-budget.mjs`:

1. **Path → group mapping** in `pr-scope-budget.json`: map changed file paths to subsystem groups (recall, extraction, config, qmd, wearables, access, graph, causal, session, maintenance, etc.) using path-prefix rules.
2. **Multi-group + multi-issue detection**: when a PR touches 3+ distinct subsystem groups AND references 2+ distinct issue numbers in its title/body, emit a WARN listing the specific groups and issues.
3. **Escalation**: FAIL when the groups share no common subsystem ancestor (e.g., "recall regression test" + "embedding backlog feature" are unrelated). The existing `scope-budget-exempt` label escape hatch applies.

## Acceptance criteria

- [ ] `pr-scope-budget.json` gains a `subsystemGroups` path-prefix mapping
- [ ] `pr-scope-budget.mjs` classifies changed files into groups and counts distinct issue references
- [ ] WARN emitted when 3+ groups AND 2+ issues detected
- [ ] FAIL emitted when groups are unrelated (no shared prefix ancestor)
- [ ] `scope-budget-exempt` label still bypasses
- [ ] Existing line-count behavior unchanged
- [ ] Unit test or dry-run demonstrates detection on a synthetic multi-group PR

## Context

- Existing: `scripts/pr-scope-budget.mjs`, `scripts/pr-scope-budget.json`, `.github/workflows/pr-scope-budget.yml`
- AGENTS.md "Cleaner PR Workflow" rule 1: "One subsystem group per PR whenever possible"
- The `scope-budget-exempt` label already exists as the escape hatch

### PR DESCRIPTION
**Pull request by joshuaswarren**
Closes #2067

## What

Extends the PR scope budget beyond raw line count with a subsystem-group / multi-issue heuristic: a diff can be small yet unfocused when several unrelated subsystems change for several different issues in one PR.

## Changes

- **`scripts/pr-scope-budget.json`** — add `subsystemGroups`, a longest-prefix map from core path prefixes to subsystem group names, covering the major subsystems (recall, extraction, config, storage, search, graph, orchestration, provenance, trust, temporal, correction, coding, connectors, chat, compounding, maintenance, peers, memory) plus package-level groups.
- **`scripts/pr-scope-budget.mjs`** — classify each *core* changed file by its longest matching prefix; extract `#<n>` issue references from PR title + body; **WARN** when distinct groups >= 3 AND distinct issues >= 2; **FAIL** when the touched groups share no package-level path ancestor (truly unrelated subsystems) AND issues >= 2. The `scope-budget-exempt` label bypasses the fail. Existing line-count behavior is unchanged.
- **`.github/workflows/pr-scope-budget.yml`** — fetch PR title + body and pass as `--pr-meta` (data only, scanned for issue refs; budget config still read from the base ref so a PR can't weaken its own budget).
- **`tests/pr-scope-budget.test.mjs`** — cover issue extraction/dedup, longest-prefix classification, shared-ancestor depth, single-group no-op (line-count unchanged), >=3 related-group WARN, unrelated-group FAIL, exempt bypass, and `subsystemGroups` config validation.

## Verification

- `node --test tests/pr-scope-budget.test.mjs` -> 17/17 pass.
- CLI smoke test: unrelated-group + 2-issue synthetic PR -> FAIL (exit 1); same with `scope-budget-exempt` -> EXEMPT (exit 0); single-group -> PASS (exit 0).
- `npm run preflight:quick` -> OK.

---

> [!NOTE]
> **Low Risk**
> Changes only CI scope-budget scripts and workflow; failures are exempt-able and multi-issue signal is skipped on pre-#2067 base evaluators.
> 
> **Overview**
> Extends the **PR scope budget** beyond core line counts with a **subsystem + multi-issue** heuristic (#2067), so small diffs that touch many areas for several issues can still warn or fail.
> 
> **`scripts/pr-scope-budget.mjs`** classifies each core changed file via longest-prefix **`subsystemGroups`**, parses `#<n>` refs from optional PR title/body (`--pr-meta`), and layers rules on top of existing thresholds: **warn** when ≥3 groups and ≥2 issues; **fail** when ≥2 groups share no `packages/<pkg>` ancestor and ≥2 issues (still bypassed by `scope-budget-exempt`). Renames count **both** core sides for grouping. Line-count and cross-scope reasons are merged into one verdict.
> 
> **`scripts/pr-scope-budget.json`** adds the large path-prefix → group map. **`.github/workflows/pr-scope-budget.yml`** fetches fresh PR text and passes `--pr-meta` only when the **base-ref** evaluator declares that flag (so older base branches do not break). **Tests** cover extraction, classification, ancestor logic, warn/fail/exempt, and rename edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit cd687837548843b170bc403d4a066abe94aea050. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Enhanced pull request scope checks now consider changed files plus the pull request title/body to detect multi-issue and multi-subsystem impact using configurable grouping.
  * Updated the scope-budget heuristic to factor shared package ancestry when deciding warn vs fail outcomes.
* **Bug Fixes**
  * Scope exemptions now consistently bypass both size-based and cross-scope failures.
  * PR scope signals refresh when PR title/body is edited.
* **Tests**
  * Added coverage for issue reference extraction/deduplication, subsystem classification, shared ancestry depth logic, threshold/group validation, and exempt behavior.

### CODE REVIEW

File: scripts/pr-scope-budget.mjs [hunk 1]
```diff
@@ -107,26 +187,63 @@ export function evaluateScopeBudget({ files, labels, thresholds, ignorePatterns
...
+  // over the same core files the line budget counts, so docs/tests/artifact
+  // churn never inflates the group count.
+  const groups = new Set();
+  const prefixes = new Set();
+  for (const file of coreFiles) {
+    const classified = classifySubsystem(file.filename, subsystemGroups);
+    if (classified) {
+      groups.add(classified.group);
+      prefixes.add(classified.prefix);
+    }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply scripts/pr-scope-budget.mjs:225]: Fixed in 8a116831: rename entries now classify by the core side (previous_filename) when the destination is non-core/ignored, so renamed core paths contribute their subsystem group. Covered by a new test.

File: scripts/pr-scope-budget.mjs [hunk 2]
```diff
@@ -107,26 +187,63 @@ export function evaluateScopeBudget({ files, labels, thresholds, ignorePatterns
     .map((f) => `${f.filename} (${f.lines})`)
     .join(", ");
 
+  // Subsystem-group / multi-issue signal (issue #2067). Classification runs
+  // over the same core files the line budget counts, so docs/tests/artifact
+  // churn never inflates the group count.
+  const groups = new Set();
+  const prefixes = new Set();
+  for (const file of coreFiles) {
+    const classified = classifySubsystem(file.filename, subsystemGroups);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply scripts/pr-scope-budget.mjs]: Fixed in 8a116831: coreFiles now carry a classifyPath = currentCounts ? filename : previous_filename, and grouping uses it — so a core file renamed out to docs still classifies (e.g. recall). New test asserts the recall+cli split fail triggers.

File: scripts/pr-scope-budget.json [hunk 3]
```diff
@@ -1,5 +1,68 @@
...
+    "packages/remnic-core/src/importers": "connectors",
+    "packages/remnic-core/src/bulk-import": "connectors",
+    "packages/remnic-core/src/chat": "chat",
+    "packages/remnic-core/src/compounding": "compounding",
+    "packages/remnic-core/src/procedural": "compounding",
+    "packages/remnic-core/src/maintenance": "maintenance",
+    "packages/remnic-core/src/peers": "peers",
+    "packages/remnic-core/src/memory": "memory",
+    "packages/remnic-core/src/cli": "core-cli",
+    "packages/remnic-cli/": "cli",
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply scripts/pr-scope-budget.json:59]: Fixed in 8a116831: added a packages/remnic-core/ catch-all group. Longest-prefix wins so fine subsystems still take precedence; verified access-mcp.ts and importance.ts now classify as remnic-core.

File: .github/workflows/pr-scope-budget.yml [hunk 4]
```diff
@@ -136,9 +144,21 @@ jobs:
...
+            --thresholds /tmp/budget-thresholds.json
+            --ignore /tmp/budget-ignore
+          )
+          # The evaluator runs from the BASE ref (trusted code). Pass the
+          # multi-issue --pr-meta signal (issue #2067) only when that base
+          # copy actually understands the flag — a base predating #2067 would
+          # otherwise reject the unknown flag during the merge-window
+          # transition. Self-heals once #2067 is on the base branch.
+          if grep -q -- '--pr-meta' scripts/pr-scope-budget.mjs; then
+            args+=(--pr-meta /tmp/pr-meta.txt)
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply .github/workflows/pr-scope-budget.yml]: Fixed in 8a116831: the gate now uses grep -qF '"--pr-meta"' to match the quoted flag literal in the parser's CLI_FLAGS declaration, not any mention in comments/usage text.

File: scripts/pr-scope-budget.mjs [hunk 5]
```diff
@@ -65,14 +103,56 @@ export function loadThresholds(thresholdsPath) {
...
+    failLines: parsed.failLines,
+    subsystemGroups: loadSubsystemGroups(parsed.subsystemGroups),
+  };
+}
+
+/** All #<n> issue references in free text, deduped (issue #2067). */
+export function extractIssueRefs(text) {
+  const issues = new Set();
+  if (typeof text !== "string") return issues;
+  for (const match of text.matchAll(/#(\d+)/g)) {
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply scripts/pr-scope-budget.mjs]: Fixed in 8a116831: extractIssueRefs now strips fenced+inline code spans and requires issue-reference boundaries (no adjacent alphanumerics or leading #), so #123456 in a code block, run#7, x#8y, and ##9 no longer count. New test covers this.

File: scripts/pr-scope-budget.mjs [hunk 6]
```diff
@@ -92,13 +179,18 @@ export function evaluateScopeBudget({ files, labels, thresholds, ignorePatterns
...
     }
     const additions = Number.isInteger(file.additions) ? file.additions : 0;
     const deletions = Number.isInteger(file.deletions) ? file.deletions : 0;
     coreLines += additions + deletions;
-    coreFiles.push({ filename, lines: additions + deletions });
+    // Classify by the CORE side of a rename: a core file renamed into a
+    // non-core/ignored destination still changed that subsystem, so grouping
+    // must follow previous_filename when the new name is not itself core
+    // (review: cursor + codex on #2067).
+    coreFiles.push({ filename, lines: additions + deletions, classifyPath: currentCounts ? filename : previous });
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply scripts/pr-scope-budget.mjs]: Fixed in 03914d4e: renames now collect every core side into classifyPaths and grouping iterates all of them, so a core-to-core move (e.g. recall -> remnic-cli) contributes both subsystems. New test asserts the two-group split fail on a single-entry core-to-core rename.

File: scripts/pr-scope-budget.mjs [hunk 7]
```diff
@@ -65,14 +103,63 @@ export function loadThresholds(thresholdsPath) {
...
+ * All #<n> issue references in free text, deduped (issue #2067). Fenced and
+ * inline code spans are dropped first so hashes inside code samples (a CSS hex
+ * color like #123456, a shell prompt) are never mistaken for issue refs, and
+ * the match requires GitHub issue-reference boundaries (no leading/trailing
+ * alphanumerics) so `abc#12`, `#12ab`, and `##12` do not count.
+ */
+export function extractIssueRefs(text) {
+  const issues = new Set();
+  if (typeof text !== "string") return issues;
+  const prose = text.replace(/```[\s\S]*?```/g, " ").replace(/`[^`]*`/g, " ");
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply scripts/pr-scope-budget.mjs]: Fixed in cd687837: extractIssueRefs now strips  HTML comments before scanning, so hidden/template comment hashes no longer count. Test extended to cover a hidden  not counting.
